Per game profiles backend #3

maniman303 opened the pull request:
This PR adds backend functionality for keeping and applying per game performance profiles. Right now allows to save per game refresh rate and fps limit.

Simple User Story:
As a user I want
- after booting my device up to have my previous settings automatically set up
- if specific game has it's own config file these settings should be automatically applied after I start the game
- if I exit the game that was using custom profile, settings from before starting the game should be reapplied
- if I have any external display connected I want profiles to be disabled and settings set to default values, so I don't accidentally mess them up

ayufan commented on PowerControl/Helpers/ProfilesController.cs:
```diff
@@ -0,0 +1,149 @@
+
+        public static void Initialize()
+        {
+            Options.RefreshRate.Instance?.SetValueChanged((_, _, newValue) =>
```
I think this is wrong. There's nothing special about FPS Limit or Refresh Rate. I would propose instead that to `MenuWithItemOptions` we add `public string? PersistenceKey { get; set; }` and set that for every property that needs to be persisted. Then we do something like this:

```csharp
public class ProfilesController
{
    private const string DefaultName = "Default";

    private ProfileSettings DefaultSettings = new ProfileSettings(DefaultName);
    private ProfileSettings? CurrentSettings;

    private System.Windows.Forms.Timer timer = new System.Windows.Forms.Timer();

    public void ProfilesController()
    {
        MenuRoot.Stack += Menu_ValueChanged;

        timer = new System.Windows.Forms.Timer();
        timer.Interval = 1000;
        timer.Tick += delegate
        {
          timer.Enabled = false;
          try { Tick(); }
          finally { timer.Enabled = true; }
        };
        timer.Start();
    }
    
    private void Menu_ValueChanged(MenuItemWithOptions options, String oldValue, String newValue)
    {
      if (options.PersistenceKey is null)
        return;
      
      CurrentSettings.Set<string>(options.PersistenceKey, newValue);
    }

    private void OnProfileChanged(String newProfile)
    {
      CurrentSettings = new ProfileSettings(newProfile);

      foreach(var options in MenuStack.Root.AllMenuItemOptions())
      {
        if (options.PersistenceKey is null)
          continue;

        var value = CurrentSettings.Get<string>(options.PersistenceKey, "");
        if (value == "")
          continue;

        // TODO: maybe for cases where settings needs to be applied are delayed
        // We instead of false pass a delay duration
        try { options.Set(value, false); }
        catch(Exception e) { TraceException("Profiles", newProfile, e); }
      }
    }
}
```

I made also a few more changes:
- We should initialize `ProfilesController`, so not static class
- It uses hook onto MenuStack
- It iterates MenuItems to set settings
- Gets rid of special write locked, and delay handling - instead if this is needed it should rather be part of the `Item.Set(,..., delay)`
- Does not include those other changes mentioned about process discovery

maniman303 replied:
I like the idea of having keys inside menu items. I used fps and refresh rate for now, because I'm developing it on my laptop and also these two options are the most wanted in profiles, so I want to first make MVP with a good base which later on can be extended.

I will implement dynamic menu options handling, but probably write lock and delay will stay as they were. Lock is there, because change to refresh rate resets fps and it's a good thing when user changes it, but it's a bad thing when program set's it. I have an idea to split ApplyValue into ApplyValue and ApplyCleanUp and former will be used if some `silent` parameter will be set, but I have to try it first.
Delay on the other hand is crucial in it's current form. You can't delay just one option - bc refresh rate, as said before, resets fps. So you have to delay them all. Also bc thread is sleeping it makes sure OSD is "frozen" during the load and user can't change anything.
If my change from ApplyValue into ApplyValue and ApplyCleanUp will succeed and after other things we can get back to this one.

And for now I won't implement the new Tick, unless there will be working version of it. I will try to figure something on my own, too.

ayufan commented on PowerControl/Menu/MenuItemWithOptions.cs:
```diff
@@ -13,7 +13,7 @@ public class MenuItemWithOptions : MenuItem
         public Func<string?>? ResetValue { get; set; }
 
-        public event Action<MenuItemWithOptions, String?, String> ValueChanged;
+        public event Action<MenuItemWithOptions, String?, String>? ValueChanged;
```
This should not be needed, as specifically this is initalised in a construct of the class.

maniman303 replied:
Sure

ayufan commented on PowerControl/Menu/MenuItemWithOptions.cs:
```diff
@@ -91,6 +91,11 @@ public void Set(String value, bool immediately = false)
             delayTimer.Enabled = true;
         }
 
+        public void SetValueChanged(Action<MenuItemWithOptions, String?, String> action)
```
This is not needed, as someone should use `+=` operator.

maniman303 replied:
+= doesn't work if you try to change it outside namespace PowerControl.Menu. But it shouldn't matter if we will move to more general approach

ayufan reviewed (changes requested):
I like the overall direction. We can just do it easier. I posted changes with some snippets how it could look like :)

maniman303 commented on PowerControl/Controller.cs:
```diff
@@ -418,6 +420,8 @@ private void DisplayChangesHandler(object? sender, EventArgs e)
                 context?.Post((object? state) =>
                 {
                     rootMenu.Update();
+                    Options.RefreshRate.Instance?.Reset();
```
Ok, VS for whatever reason thinks it can be.

maniman303 commented on PowerControl/Helpers/DeviceManager.cs:
```diff
@@ -35,6 +35,16 @@ public static bool RefreshDisplays()
+                return false;
+            }
+
+            return screens[0].Bounds.Size.Width == 1280 && screens[0].Bounds.Size.Height == 800;
```
I will try to get max possible resolution for the screen, I believe this is already done in resolution menu.

ayufan replied:
I would probably get a list of displays (all, not only connected) and assume that the name of first one is our internal display.

maniman303 replied:
Sounds fair

ayufan commented on CommonHelpers/RTSS.cs:
```diff
@@ -76,6 +76,26 @@ public static bool SetProfileProperty<T>(string propertyName, T value)
             }
         }
 
+        public static string? GetCurrentGameName()
```
I don't know yet how Default should work yet. I would consider to:

- initially ignore default settings - we can get back to it next
- we can force profile apply on display changed
- we can even store profile settings per display name - and apply them on a foreground change, and display change

maniman303 replied:
I do not agree with supporting multiple displays in a single profile. It's not user friendly, it's complex and that's where I draw a line. And without default settings for the deck display we can't restore them after we exit the game, which would be the opposite of the whole idea.

maniman303 commented on PowerControl/Helpers/ProfileSettings.cs:
```diff
@@ -0,0 +1,53 @@
+        public ProfileSettings(string profileName) : base("Profile")
+        {
+            this.TouchSettings = true;
+            this.ConfigFile = Path.Combine(profilesPath, profileName + ".ini");
```
Ok, will use, missed that one. We already ensure directory exists in static constructor of ProfilesSettings. And I believe it's enough (but I can add it to the Set), bc if we start adding the check to Set, maybe we should add it to Get, too? Maybe we should add it to Exists? And we will end up with a dozen checks

maniman303 commented on PowerControl/Helpers/ProfileSettings.cs:
```diff
@@ -0,0 +1,53 @@
+            return base.Set(key, value);
+        }
+
+        public static bool CheckIfExists(string profileName)
```
Sounds good!

maniman303 commented:
@ayufan I have implemented most of the review changes and also I've added simple User Story in the description so it's easier for you to understand my idea. I have not implemented yours idea for detecting and handling profiles changes as in my opinion it's too broken and complex to current solution that gets the job done.

Although I'm eager to explore the idea of using the profile of the current FOREGROUND game, even if other games are running in the background, especially as it can fix some rare issues when `OSD.GetAppEntries(AppFlags.MASK)` could return more than one entry, even if there's only one game? But first you would have to tell me how to reproduce an issue like that (without using another game) and explain to me what's the function of `RTSS.IsOSDForeground()`.

I will play around a bit with `RTSS.IsOSDForeground(out int processId)`, maybe extend it to return application name.

ayufan commented on PowerControl/Helpers/ProfileSettings.cs:
```diff
@@ -0,0 +1,42 @@
+
+        static ProfileSettings()
+        {
+            Directory.CreateDirectory(profilesPath);
```
This is not best, since it assumes that CWD equals to the current process directory. Which is mostly true, but not always.

ayufan reviewed (commented):
Overall this goes into good direction. I will merge this in bits reworking some aspects.

- I'm unsure about the interface changes to `AfterApply` and `Set` and I still need to understand fully `ProfilesController` behavior.
- I'm still not convinced about Default behavior. I need to re-consider this, but something makes me feel that this can be done easier. I was thinking that maybe all changes made by profile are snapshotted, and then restored to prior state.
- I still have a conceptual problem with `AfterApply` to reset `FPSLimit` setting.
  - Maybe instead properties should define an order of execution to ensure that settings are enabled in correct order, then we would not need AfterApply. This could be defined as: that this setting needs to be applied after those settings.
  - Maybe settings would define what other settings they do reset, and this way the execution order would be evaluated - creating DAG.
- I also wonder about `TroubledKey` where we apply the delay. Maybe this is a setting that should be put in `.ini` - now maybe manually, but later with some configurable setting.

ayufan commented:
@maniman303 This is merged into `main`: https://github.com/ayufan/steam-deck-tools/commit/3252e799cb298edb3237cc86b70d975406d06f6f. I think original kudos goes to you doing this. I reworked next some bits to tidy this up, and also be able to track many processes, and did add OSD as well.

Thank you for spending time on this.
